tpu-provisioner/internal/cloud/gke.go

```diff
@@ -340,23 +352,24 @@ func sumTPURequests(p *corev1.Pod) (int, error) {
 	return n, nil
 }
 
-func podToNodePoolName(p *corev1.Pod, prefix, suffix string) string {
-	// Use the UID of the Pod's owner (falling back to the Pod UID if it has
-	// no owner) as the unique identifier for the node pool.
-	// It is necessary to use something that is unique to the Pod not the Job/JobSet
-	// because the scheduler is not guaranteed to place Pods on the same
-	// node pools that were created for them. This commonly happens when
-	// node pools are reused by other Jobs after the original Job has completed
-	// or restarted. Using another identifier like the job-key could result in
-	// deadlocks in this case.
-	var uid string
-	ref := metav1.GetControllerOf(p)
-	if ref != nil {
-		uid = string(ref.UID)
-	} else {
-		uid = string(p.UID)
-	}
-	return prefix + uid[0:12] + suffix
+// podToNodePoolName deterministically generates a node pool name for a given pod,
+// by using the JobSet name and job-key (SHA1 hash of namespaced job key), as
+// given in the pod labels.
+// These labels are stable through JobSet restarts, so the node pool name
+// generated here will be the same if the JobSet is restarted.
+// Node pool name format is: {first 34 chars of jobset name}-{first 5 chars of job-key}
+// This ensures node pool names are within the 40 char limit on node pool name size.
+func podToNodePoolName(p *corev1.Pod) (string, error) {
+	jobSetName, exists := p.Labels[jobset.JobSetNameKey]
+	if !exists {
+		return "", fmt.Errorf("%s label not found on pod %s", jobset.JobSetNameKey, p.Name)
+	}
+	jobKey, exists := p.Labels[jobset.JobKey]
+	if !exists {
+		return "", fmt.Errorf("%s label not found on pod %s", jobset.JobKey, p.Name)
+	}
+	nodePoolName := fmt.Sprintf("%s-%s", jobSetName[:34], jobKey[:5])
```

Comment:
`:34` will panic if the name is not that long

Reply:
Ah I was assuming slicing in Go worked the same way as in Python, I guess not! Thanks, fixed.